File: 백준/Silver/9012. 괄호/괄호.c
#include <stdio.h>
#include <string.h>

int main(void) {  
   int a, count, flag;
   scanf("%d", &a);

   char arr[51];
   for(int i = 0; i < a; i++){
      flag = 1;
      count = 0;
      scanf("%s", arr);
      int length = strlen(arr);
      for(int j = 0; j < length; j++){
         if(arr[j] == '(' ){
            count += 1;
         }
         else{
            count -= 1;
         }
         if(count < 0){
            flag = 0;
            break;
         }
      }
      if(count != 0 || !flag){
         printf("NO\n");
      }
      else if(count == 0 && arr[length - 1] == ')'){
         printf("YES\n");
      }
   }

   return 0;
}

File: 백준/Silver/9012. 괄호/괄호.py
a = int(input())
for i in range(a):
  count = 0
  flag = True
  arr = input()
  for j in range(len(arr)):
    if arr[j] == '(':
      count += 1
    else:
      count -= 1

    if count < 0:
      flag = False
      break

  if (not flag or count != 0):
    print("NO")
  elif (arr[len(arr) - 1] == ')' and count == 0):
    print("YES")